File: src/main/java/no/novari/flyt/archive/gateway/template/model/CollectionTemplate.java
package no.novari.flyt.archive.gateway.template.model;

import com.fasterxml.jackson.annotation.JsonInclude;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.extern.jackson.Jacksonized;

@Getter
@Builder
@EqualsAndHashCode
@Jacksonized
@JsonInclude(JsonInclude.Include.NON_NULL)
public class CollectionTemplate<T> {

    @NotNull
    @Valid
    private final T elementTemplate;

}

File: src/main/kotlin/no/novari/flyt/archive/gateway/template/model/CollectionTemplate.kt
package no.novari.flyt.archive.gateway.template.model

import com.fasterxml.jackson.annotation.JsonInclude
import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull

@JsonInclude(JsonInclude.Include.NON_NULL)
data class CollectionTemplate<T>(
    @field:NotNull
    @field:Valid
    val elementTemplate: T,
) {
    companion object {
        @JvmStatic
        fun <T> builder() = Builder<T>()
    }

    class Builder<T> {
        private var elementTemplate: T? = null

        fun elementTemplate(elementTemplate: T) = apply { this.elementTemplate = elementTemplate }

        fun build() = CollectionTemplate(requireNotNull(elementTemplate) { "elementTemplate is required" })
    }
}
